token/js/src/extensions/tokenMetadata/actions.ts

```diff
@@ -0,0 +1,390 @@
+import type { ConfirmOptions, Connection, PublicKey, Signer, TransactionSignature } from '@solana/web3.js';
+import type { Field, TokenMetadata } from '@solana/spl-token-metadata';
+import {
+    createEmitInstruction,
+    createInitializeInstruction,
+    createRemoveKeyInstruction,
+    createUpdateAuthorityInstruction,
+    createUpdateFieldInstruction,
+    pack,
+} from '@solana/spl-token-metadata';
+import { sendAndConfirmTransaction, SystemProgram, Transaction } from '@solana/web3.js';
+import { TokenAccountNotFoundError } from '../../errors.js';
+
+import { TOKEN_2022_PROGRAM_ID } from '../../constants.js';
+import { getSigners } from '../../actions/internal.js';
+import { unpackMint } from '../../state/mint.js';
+import { getExtensionData, ExtensionType } from '../extensionType.js';
+
+/**
+ * Calculates additional lamports need to variable length extension
+ *
+ * @param connection       Connection to use
+ * @param address          Mint Account
+ * @param tokenMetadata    Token Metadata
+ * @param programId        SPL Token program account
+ *
+ * @return lamports to send
+ */
+async function getAdditionalRentForNewMetadata(
```

Comment:
Do you have any suggestions for how to implement this? I don't think the implementation mimics very well how the rust code does it, its not very generic - and it doesn't have the complexity of handling updates yet.


For updates, the only way i can think of is basically writing "reducer" function like `(metadata, updates) => newMetadata` and then getting calculating the length on this new Metadata. I think its relatively straightforward to implemented but there are a few edge cases like handling Fields and adding/removing/changing additional metadata).

Wanted to check if there is a better way, before i go and do this cause it feels like we are re-implementing the rust code.

Reply:
You'll have to make this function take a `connection: Connection` arg and then do exactly what the Rust client does.

First get the mint with `connection.getAccountInfo(..)`, check the new metadata size against the existing one to see if more rent is needed, and if so use `connection.getMinimumBalanceForRentExemption(..)` to get those additional required lamports.

Reply:
You can possibly piggyback on the `getTokenMetadata` which you've written already

Reply:
Okay i've gotten something that works. I tried to make it as simple as possible, but still feels quite complicated. Let me know if you think theres anything that can be simplified

Reply:
Yeah, it does seem a little bit complicated. I think we can tighten up some of the null/error handling into fewer lines, and then we can simplify the size comparison.

I left a comment below.